classes/base.lua

```diff
@@ -182,24 +182,133 @@ function base:runHooks (category, args)
+    end, options.help, SILE.currentlyProcessingFile)
+  end, "Define a new macro. \\define[command=example]{ ... \\process }")
+
+  -- A utility function that allows SILE.call() to be used as a noop wrapper.
+  self:registerCommand("noop", function (_, content)
+    SILE.process(content)
+  end)
+
+  -- The document (SIL) or sile (XML) command is always the sigular leaf at the
+  -- top level of our AST. The work you might expect to see happen here is
+  -- actually handled by SILE.inputter:classInit() before we get here, so these
+  -- are just pass through functions. Theoretically, this could be a useful
+  -- point to hook into-especially for included documents.
+  self:registerCommand("document", function (_, content)
+    SILE.process(content)
+  end)
+  self:registerCommand("sile", function (_, content)
+    SILE.process(content)
+  end)
+
+  self:registerCommand("comment", function (_, _)
+  end, "Ignores any text within this command's body.")
+
+  self:registerCommand("process", function ()
+    SU.error("Encountered unsubstituted \\process.")
+  end, "Within a macro definition, processes the contents of the macro body.")
+
+  self:registerCommand("script", function (options, content)
     if options.src then
-      local pack, _ = require(options.src)
-      self:initPackage(pack)
+      SILE.require(options.src)
     else
-      local func, err = load(content[1])
-      if not func then SU.error(err) end
-      func()
+      SILE.processString(content[1], options.format or "lua")
     end
-  end, "Runs lua code. The code may be supplied either inline or using the src=... option. (Think HTML.)")
+  end, "Runs lua code. The code may be supplied either inline or using src=...")
 
-  SILE.registerCommand("include", function (options, _)
-      SILE.readFile(options["src"])
-  end, "Includes a SILE file for processing.")
+  self:registerCommand("include", function (options, content)
+    if options.src then
+      SILE.processFile(options.src, options.format)
+    else
+      SILE.processString(content[1], options.format)
+    end
+  end, "Includes a content file for processing.")
+
+  self:registerCommand("lua", function (options, content)
+    if options.module then
+      SILE.require(options.module)
+    elseif options.src then
+      SILE.processFile(options.src, "lua")
+    else
+      SILE.processString(content[1], "lua")
+    end
+  end, "Run Lua code. The code may be supplied either inline or using src=...")
+
+  self:registerCommand("sil", function (options, content)
+    if options.src then
+      SILE.processFile(options.src, "sil")
+    else
+      SILE.processString(content[1], "sil")
+    end
+  end, "Process sil content. The content may be supplied either inline or using src=...")
+
+  self:registerCommand("xml", function (options, content)
+    if options.src then
+      SILE.processFile(options.src, "xml")
+    else
+      SILE.processString(content[1], "xml")
```

Comment:
Nope, raw XML.

You can now nest XML in SIL or SIL in XML or Lua in SIL or XML (ala what `\script` has been). This just generalizes that to be a 3 way interchangeability rather than separate hacks for each (`doTexlike` from Lua, `\include` from XML but only as an external file, etc.). You can also programmatically pass strings to be parsed as content from any of those three languages. Lua inputs can also return an AST directly.

Reply:
Caveat: the embeded SIL parser is limited to proper closing tags. For example if you open with `\begin{xml}` the next instance of `\end{xml}` will terminate the XML embedding even if it happened to be content of the XML. For `\xml{` the next *matching* `}` will close. Both cases can be worked around by including balancing tags that are not desired as content as comments.

Reply:
Ah I see indeed, you made `lua`, `sil` and `xml` all reserved in the passthroughCommands. I am unsure how it relates to include/preamble semantics, but do as you deem fit. (I don't really get why one would want to mix xml with sil, beyond what would have been achievable with `\raw` + handler, but I'm just missing the picture. Question answered.)